libqtile/widget/mpdwidget.py

```diff
@@ -30,285 +30,270 @@
-                    playing = pangocffi.markup_escape_text(text)
+    def update_status(self, is_retry=False):
+        try:
+            self.client.command_list_ok_begin()
+            self.client.status()
+            self.client.currentsong()
+            status, current_song = self.client.command_list_end()
+            return self.formatter(status, current_song)
+        except (ProtocolError, CommandListError):
+            if is_retry:
+                return self.no_connection
             else:
-                playing = self.do_format(self.fmt_stopped)
+                return self.update_status(is_retry=True)
 
-        except Exception:
-            logger.exception('Mpd error on update')
+    def formatter(self, status, currentsong):
+        play_status = self.play_states[status['state']]
 
-        return playing
+        if currentsong:
+            # Dirty hack to prevent keys conflict
+            currentsong['fulltime'] = currentsong.get('time', '0')
+            del currentsong['time']
 
-    def poll(self):
-        was_connected = self.connected
+            status.update(currentsong)
+        self.prepare_formatting(status)
 
-        if not self.connected:
-            if self.connected is None or self.reconnect:
-                while not self.stop and not self.connect(quiet=True):
-                    time.sleep(self.reconnect_interval)
-            else:
-                return
+        if status['state'] == 'stop':
+            fmt = self.status_format_stopped
+        else:
+            fmt = self.status_format
 
-        if self.stop:
-            return True
+        try:
+            text = fmt.format(play_status=play_status, **status)
+        except KeyError as e:
+            logger.exception(f"mpd client did not return status: {e.args[0]}")
+            return "ERROR"
 
-        if was_connected:
+        if self.do_color_progress and status['state'] != 'stop':
             try:
-                self.client.ping()
-            except mpd.ConnectionError:
-                self.client.disconnect()
-                self.connected = False
-                return self.msg_nc
-            except Exception:
-                logger.exception('Error communicating with mpd')
-                self.client.disconnect()
-                return
-
-        return self._get_status()
-
-    def button_press(self, x, y, button):
-        if not self.connect():
-            return False
+                elapsed, total = status['time'].split(':')
+                percent = float(elapsed) / float(total)
+                progress = int(percent * len(text))
+                text = '<span color="{0}">{1}</span>{2}'.format(
+                    utils.hex(self.foreground_progress),
+                    pangocffi.markup_escape_text(text[:progress]),
+                    pangocffi.markup_escape_text(text[progress:]),
+                )
+            except (ZeroDivisionError, ValueError):
+                pass
+
+        return text
```

Comment:
@MagusLucius The logic for the color-changing is here (in green). Could this not follow the equivalent part on your branch at [L335](https://github.com/MagusLucius/qtile/blob/patch-1/libqtile/widget/mpd2widget.py#L335)?

Reply:
Yes, adding that would be much appreciated :)